{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%load_ext cythonmagic"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%cython\n",
      "from libc.math cimport exp, abs, sqrt, ceil\n",
      "from numpy cimport ndarray, double_t, int_t\n",
      "import numpy, cython, time, gc, random\n",
      "\n",
      "BOLTZMANN_K = 0.0019872041\n",
      "\n",
      "#ctypedef int_t   int\n",
      "#ctypedef double_t double\n",
      "\n",
      "class NucCythonError(Exception):\n",
      "\n",
      "  def __init__(self, err):\n",
      "\n",
      "    Exception.__init__(self, err)\n",
      "\n",
      "\n",
      "cdef int getRepulsionList(ndarray[int,   ndim=2] repList,\n",
      "                          ndarray[double, ndim=2] coords,\n",
      "                          int nCoords, int nRepMax,\n",
      "                          double repDist,\n",
      "                          ndarray[double, ndim=1] radii):\n",
      "  \n",
      "  cdef int i, j \n",
      "  cdef int n = 0\n",
      "  cdef double dx, dy, dz, d2\n",
      "  cdef double distLim\n",
      "  cdef double distLim2 \n",
      "  \n",
      "  for i in range(nCoords-2):\n",
      "    for j in range(i+2, nCoords):\n",
      "      distLim = repDist + radii[i] + radii[j]\n",
      "      distLim2 = distLim * distLim\n",
      "      \n",
      "      dx = coords[i,0] - coords[j,0]\n",
      "      if abs(dx) > distLim:\n",
      "        continue\n",
      "\n",
      "      dy = coords[i,1] - coords[j,1]\n",
      "      if abs(dy) > distLim:\n",
      "        continue\n",
      "\n",
      "      dz = coords[i,2] - coords[j,2]\n",
      "      if abs(dz) > distLim:\n",
      "        continue\n",
      "\n",
      "      d2 = dx*dx + dy*dy + dz*dz\n",
      " \n",
      "      if d2 > distLim2:\n",
      "        continue\n",
      "\n",
      "      repList[n,0] = i\n",
      "      repList[n,1] = j\n",
      "      \n",
      "      n += 1\n",
      "      \n",
      "      #return if we reached max, something is seriously wrong anyway\n",
      "      if n == nRepMax:\n",
      "        return n\n",
      "      \n",
      "  return n\n",
      "\n",
      "\n",
      "cdef double getTemp(ndarray[double] masses,\n",
      "                    ndarray[double, ndim=2] veloc,\n",
      "                    int nCoords): \n",
      "  cdef int i\n",
      "  cdef double kin = 0.0\n",
      "  \n",
      "  for i in range(nCoords):\n",
      "    kin += masses[i] * (veloc[i,0]*veloc[i,0] + veloc[i,1]*veloc[i,1] + veloc[i,2]*veloc[i,2])\n",
      "\n",
      "  return kin / (3 * nCoords * BOLTZMANN_K)\n",
      "  \n",
      "\n",
      "def getStats(ndarray[int,   ndim=2] restIndices,\n",
      "             ndarray[double, ndim=2] restLimits,\n",
      "             ndarray[double, ndim=2] coords,\n",
      "             int nRest):      \n",
      "\n",
      "  cdef int i, nViol = 0\n",
      "  cdef int j, k\n",
      "  cdef double viol, dmin, dmax, dx, dy, dz, r, s = 0\n",
      "  \n",
      "  for i in range(nRest):\n",
      "    j = restIndices[i,0]\n",
      "    k = restIndices[i,1]\n",
      "    \n",
      "    if j == k:\n",
      "      continue\n",
      "\n",
      "    dmin = restLimits[i,0]\n",
      "    dmax = restLimits[i,1]\n",
      "    \n",
      "    dx = coords[j,0] - coords[k,0]\n",
      "    dy = coords[j,1] - coords[k,1]\n",
      "    dz = coords[j,2] - coords[k,2]\n",
      "    r = sqrt(dx*dx + dy*dy + dz*dz)\n",
      "\n",
      "    if r < dmin:\n",
      "      viol = dmin - r\n",
      "      nViol += 1\n",
      "      \n",
      "    elif r > dmax:\n",
      "      viol = r - dmax\n",
      "      nViol += 1\n",
      "      \n",
      "    else:\n",
      "      viol = 0\n",
      "\n",
      "    s += viol * viol\n",
      "\n",
      "  return nViol, sqrt(s/nRest)\n",
      "\n",
      "\n",
      "cdef void updateMotion(ndarray[double] masses,\n",
      "                       ndarray[double, ndim=2] forces,\n",
      "                       ndarray[double, ndim=2] accel,\n",
      "                       ndarray[double, ndim=2] veloc,\n",
      "                       ndarray[double, ndim=2] coords,\n",
      "                       int nCoords, double tRef,\n",
      "                       double tStep, double beta):\n",
      "\n",
      "  cdef int i\n",
      "  cdef double r, rtStep, temp\n",
      "\n",
      "  rtStep = 0.5 * tStep * tStep\n",
      "  temp = getTemp(masses, veloc, nCoords)\n",
      "  temp = max(temp, 0.001)\n",
      "  r = beta * (tRef/temp-1.0)\n",
      "  \n",
      "  for i in range(nCoords):\n",
      " \n",
      "    accel[i,0] = forces[i,0] / masses[i] + r * veloc[i,0]\n",
      "    accel[i,1] = forces[i,1] / masses[i] + r * veloc[i,1]\n",
      "    accel[i,2] = forces[i,2] / masses[i] + r * veloc[i,2]\n",
      "\n",
      "    coords[i,0] += tStep * veloc[i,0] + rtStep * accel[i,0]\n",
      "    coords[i,1] += tStep * veloc[i,1] + rtStep * accel[i,1]\n",
      "    coords[i,2] += tStep * veloc[i,2] + rtStep * accel[i,2]\n",
      " \n",
      "    veloc[i,0] += tStep * accel[i,0]\n",
      "    veloc[i,1] += tStep * accel[i,1]\n",
      "    veloc[i,2] += tStep * accel[i,2]\n",
      "\n",
      "\n",
      "cdef void updateVelocity(ndarray[double] masses,\n",
      "                         ndarray[double, ndim=2] forces,\n",
      "                         ndarray[double, ndim=2] accel,\n",
      "                         ndarray[double, ndim=2] veloc,\n",
      "                         int nCoords, double tRef,\n",
      "                         double tStep, double beta):\n",
      "\n",
      "  cdef int i\n",
      "  cdef double r, temp\n",
      "\n",
      "  temp = getTemp(masses, veloc, nCoords)\n",
      "  #avoid division by 0 temperature\n",
      "  temp = max(temp, 0.001)\n",
      "  r = beta * (tRef/temp-1.0)\n",
      "\n",
      "  for i in range(nCoords):\n",
      "    veloc[i,0] += 0.5 * tStep * (forces[i,0] / masses[i] + r * veloc[i,0] - accel[i,0])\n",
      "    veloc[i,1] += 0.5 * tStep * (forces[i,1] / masses[i] + r * veloc[i,1] - accel[i,1])\n",
      "    veloc[i,2] += 0.5 * tStep * (forces[i,2] / masses[i] + r * veloc[i,2] - accel[i,2])\n",
      "\n",
      "\n",
      "cdef double getRepulsiveForce(ndarray[int,   ndim=2] repList,\n",
      "                              ndarray[double, ndim=2] forces,\n",
      "                              ndarray[double, ndim=2] coords,\n",
      "                              int nRep, double fConst,\n",
      "                              ndarray[double, ndim=1] radii):\n",
      "\n",
      "  cdef int i, j, k\n",
      "  cdef double dx, dy, dz, d2, dr, rjk \n",
      "  cdef double force = 0\n",
      "  cdef double repDist2 \n",
      "\n",
      "  if fConst == 0:\n",
      "    return force\n",
      "\n",
      "  for i from 0 <= i < nRep:\n",
      "    j = repList[i,0]\n",
      "    k = repList[i,1]\n",
      "    repDist = radii[j] + radii[k]\n",
      "    repDist2 = repDist * repDist\n",
      "\n",
      "\n",
      "    dx = coords[k,0] - coords[j,0]\n",
      "    if abs(dx) > repDist:\n",
      "      continue\n",
      "\n",
      "    dy = coords[k,1] - coords[j,1]\n",
      "    if abs(dy) > repDist:\n",
      "      continue\n",
      "\n",
      "    dz = coords[k,2] - coords[j,2]\n",
      "    if abs(dz) > repDist:\n",
      "      continue\n",
      "\n",
      "    d2 = dx*dx + dy*dy + dz*dz\n",
      "    if d2 > repDist2:\n",
      "      continue\n",
      "\n",
      "    dr = repDist2 - d2\n",
      "    #energy contribution\n",
      "    force += fConst * dr * dr\n",
      "    rjk = 4 * fConst * dr\n",
      "\n",
      "    dx *= rjk\n",
      "    dy *= rjk\n",
      "    dz *= rjk\n",
      "\n",
      "    #force contributions\n",
      "    forces[j,0] -= dx\n",
      "    forces[k,0] += dx\n",
      "\n",
      "    forces[j,1] -= dy\n",
      "    forces[k,1] += dy\n",
      "\n",
      "    forces[j,2] -= dz\n",
      "    forces[k,2] += dz\n",
      "      \n",
      "  return force\n",
      "    \n",
      "\n",
      "cdef double getRestraintForce(ndarray[double, ndim=2] forces,\n",
      "                              ndarray[double, ndim=2] coords,\n",
      "                              ndarray[int,   ndim=2] restIndices,\n",
      "                              ndarray[double, ndim=2] restLimits,\n",
      "                              ndarray[int, ndim=1] restAmbig,\n",
      "                              int nRest, double fConst, double exponent=2.0,\n",
      "                              double distSwitch=0.5, double asymptote=1.0):\n",
      "               \n",
      "  cdef int i, j, k, n, nAmbig\n",
      "  cdef double a, b, da, d, dmin, dmax, dx, dy, dz\n",
      "  cdef double r, r2, s2, rjk, ujk, force = 0, t\n",
      "\n",
      "  b = asymptote*distSwitch*distSwitch - exponent*distSwitch*distSwitch*distSwitch\n",
      "  a = distSwitch*distSwitch - asymptote*distSwitch - b/ distSwitch\n",
      "\n",
      "  i = 0\n",
      "  \n",
      "  while i < nRest:\n",
      "\n",
      "\n",
      "    nAmbig = restAmbig[i]\n",
      "    r2 = 0.0\n",
      "    \n",
      "    if nAmbig == 1:\n",
      "      j = restIndices[i,0]\n",
      "      k = restIndices[i,1]\n",
      "      \n",
      "      if j != k:\n",
      "        dx = coords[j,0] - coords[k,0]\n",
      "        dy = coords[j,1] - coords[k,1]\n",
      "        dz = coords[j,2] - coords[k,2]\n",
      "        r2 = dx*dx + dy*dy + dz*dz\n",
      "    \n",
      "    else:\n",
      "      for n in range(nAmbig):\n",
      "        j = restIndices[i+n,0]\n",
      "        k = restIndices[i+n,1]\n",
      "        \n",
      "        if j != k:\n",
      "          dx = coords[j,0] - coords[k,0]\n",
      "          dy = coords[j,1] - coords[k,1]\n",
      "          dz = coords[j,2] - coords[k,2]\n",
      "          r = dx*dx + dy*dy + dz*dz\n",
      "          r2 += 1.0 / (r * r)\n",
      "\n",
      "      if r2 > 0:\n",
      "        r2 = 1.0 / sqrt(r2)\n",
      "\n",
      "    if r2 <= 0:\n",
      "      i += nAmbig\n",
      "      continue\n",
      "\n",
      "    dmin = restLimits[i,0]\n",
      "    dmax = restLimits[i,1]\n",
      "\n",
      "    da = dmax + distSwitch\n",
      "\n",
      "    if r2 < dmin*dmin:\n",
      "      r2 = max(r2, 1e-8)\n",
      "      r = sqrt(r2)\n",
      "      d = dmin - r\n",
      "      ujk = fConst * d * d\n",
      "      rjk = fConst * 2 * d\n",
      "    \n",
      "    elif r2 > dmax*dmax:\n",
      "      r = sqrt(r2)\n",
      "      d = r - dmax\n",
      "\n",
      "      if r <= da:\n",
      "        ujk = fConst * d * d\n",
      "        rjk = - fConst * 2 * d\n",
      "\n",
      "      else:\n",
      "        ujk = fConst * (a + asymptote*d + b/d)\n",
      "        rjk = - fConst * (asymptote - b/(d*d))\n",
      "\n",
      "    else:\n",
      "      ujk = rjk = 0\n",
      "      r = 1.0\n",
      "\n",
      "    force += ujk\n",
      "\n",
      "    if nAmbig == 1:\n",
      "      j = restIndices[i,0]\n",
      "      k = restIndices[i,1]\n",
      "\n",
      "      if j == k:\n",
      "        i += nAmbig\n",
      "        continue\n",
      "      \n",
      "      t = rjk / r\n",
      "      dx = coords[j,0] - coords[k,0]\n",
      "      dy = coords[j,1] - coords[k,1]\n",
      "      dz = coords[j,2] - coords[k,2]\n",
      "      \n",
      "      dx *= t\n",
      "      dy *= t\n",
      "      dz *= t\n",
      "      \n",
      "      forces[j,0] += dx\n",
      "      forces[j,1] += dy\n",
      "      forces[j,2] += dz\n",
      "      \n",
      "      forces[k,0] -= dx\n",
      "      forces[k,1] -= dy\n",
      "      forces[k,2] -= dz\n",
      "\n",
      "    else:\n",
      "\n",
      "      for n in range(nAmbig):\n",
      "        j = restIndices[i+n,0]\n",
      "        k = restIndices[i+n,1]\n",
      "        \n",
      "        if j == k:\n",
      "          continue\n",
      "          \n",
      "        dx = coords[j,0] - coords[k,0]\n",
      "        dy = coords[j,1] - coords[k,1]\n",
      "        dz = coords[j,2] - coords[k,2]\n",
      "        \n",
      "        s2 = dx*dx + dy*dy + dz*dz\n",
      "        t = rjk * r2 * r2 * r / (s2 * s2 * s2)\n",
      "        \n",
      "        dx *= t\n",
      "        dy *= t\n",
      "        dz *= t\n",
      "        \n",
      "        forces[j,0] += dx\n",
      "        forces[k,0] -= dx\n",
      "        forces[j,1] += dy\n",
      "        forces[k,1] -= dy\n",
      "        forces[j,2] += dz\n",
      "        forces[k,2] -= dz\n",
      "      \n",
      "    i += nAmbig\n",
      "\n",
      "  return force\n",
      "\n",
      "\n",
      "def getSupportedPairs(ndarray[int, ndim=2] positions,\n",
      "                     int threshold=2000000, int posErr=100):\n",
      "                  \n",
      "  cdef int i, j, n = len(positions)\n",
      "  cdef int pA, pB, pC, pD\n",
      "  cdef ndarray[int, ndim=1] supported = numpy.zeros(n, numpy.int32)\n",
      "  \n",
      "  for i in range(n):\n",
      "    pA = positions[i,0]\n",
      "    pB = positions[i,1]\n",
      "\n",
      "    for j in range(n):\n",
      "      if j == i:\n",
      "        continue\n",
      "      \n",
      "      pC = positions[j,0]\n",
      "      pD = positions[j,1]\n",
      "      \n",
      "      if (posErr < abs(pC-pA) < threshold) and (posErr < abs(pD-pB) < threshold):\n",
      "        supported[i] = 1\n",
      "        break\n",
      " \n",
      "      elif (posErr < abs(pD-pA) < threshold) and (posErr < abs(pC-pB) < threshold):\n",
      "        supported[i] = 1\n",
      "        break\n",
      "    \n",
      "  indices = supported.nonzero()[0]\n",
      "  \n",
      "  return indices\n",
      "\n",
      "\n",
      "def getInterpolatedCoords(chromosomes, posDict, prevPosDict, ndarray[double, ndim=2] coords):\n",
      "  \"\"\"\n",
      "  Interpolate x,y,z particle positions for an array of seq positions to a new  \n",
      "  seq positions e.g. for a change in binned resolution.\n",
      "  \"\"\"\n",
      "  \n",
      "  cdef int i, j, i0, j0, p1, p2, n, m, d, dMin\n",
      "  cdef double f, g\n",
      "  cdef ndarray[int, ndim=1] positions\n",
      "  cdef ndarray[int, ndim=1] prevPositions\n",
      "  \n",
      "  cdef int a, b\n",
      "  \n",
      "  # Get index offsets for each chromo  \n",
      "  i = 0\n",
      "  offsets = {}\n",
      "  for chrA in chromosomes:\n",
      "    offsets[chrA] = i\n",
      "    i += len(posDict[chrA])\n",
      "  \n",
      "  a = i\n",
      "  cdef ndarray[double, ndim=2] newCoords = numpy.empty((i, 3), float)\n",
      "  \n",
      "  i = 0\n",
      "  prevOffsets = {}\n",
      "  for chrA in chromosomes:\n",
      "    prevOffsets[chrA] = i\n",
      "    i += len(prevPosDict[chrA])  \n",
      "  \n",
      "  b = i\n",
      "  \n",
      "  for chrA in posDict:\n",
      "    i0 = offsets[chrA]\n",
      "    j0 = prevOffsets[chrA]\n",
      "    positions = posDict[chrA]\n",
      "    prevPositions = prevPosDict[chrA]\n",
      "    n = len(positions)\n",
      "    m = len(prevPositions)\n",
      "    \n",
      "    for i in range(n):\n",
      "\n",
      "      #find closest old positions for coordinate interpolation\n",
      "      p1 = 0\n",
      "      dMin = positions[i]-prevPositions[0]\n",
      "      \n",
      "      for j in range(1,m):\n",
      "        d = positions[i]-prevPositions[j]\n",
      "        \n",
      "        if abs(d) < abs(dMin):\n",
      "          p1 = j\n",
      "          dMin = d #closest pos\n",
      "          \n",
      "        elif abs(d) > abs(dMin): # Seq positions were in order\n",
      "          break  \n",
      "    \n",
      "      if dMin == 0: #new position coincides with an old position\n",
      "        p2 = p1\n",
      "     \n",
      "      elif dMin > 0: #new pos is above p1\n",
      "        p2 = min(p1+1, m-1)\n",
      "      \n",
      "      else: #new pos is below p1\n",
      "        p2 = p1\n",
      "        p1 = max(0, p1-1)\n",
      "        dMin = positions[i] - prevPositions[p1]\n",
      "      \n",
      "      #calculate coordinates\n",
      "      if prevPositions[p2] == prevPositions[p1]:\n",
      "        newCoords[i0+i, 0] = coords[j0+p1, 0]\n",
      "        newCoords[i0+i, 1] = coords[j0+p1, 1]\n",
      "        newCoords[i0+i, 2] = coords[j0+p1, 2]\n",
      " \n",
      "      else: #interpolate\n",
      "        f = <float>dMin/<float>(prevPositions[p2]-prevPositions[p1])\n",
      "        g = 1.0 - f\n",
      "        \n",
      "        newCoords[i0+i, 0] = g * coords[j0+p1, 0] + f * coords[j0+p2, 0]\n",
      "        newCoords[i0+i, 1] = g * coords[j0+p1, 1] + f * coords[j0+p2, 1]\n",
      "        newCoords[i0+i, 2] = g * coords[j0+p1, 2] + f * coords[j0+p2, 2]\n",
      "  \n",
      "  return newCoords\n",
      "  \n",
      "\n",
      "\n",
      "def runDynamics(ndarray[double, ndim=2] coords,\n",
      "                ndarray[double, ndim=1] masses,\n",
      "                ndarray[double, ndim=1] radii,\n",
      "                ndarray[int, ndim=2] restIndices,\n",
      "                ndarray[double, ndim=2] restLimits,\n",
      "                ndarray[int, ndim=1] restAmbig,\n",
      "                double tRef=1000.0, double tStep=0.001, int nSteps=1000,\n",
      "                double fConstR=1.0, double fConstD=25.0, double beta=10.0,\n",
      "                double repDist=1.5, # A heuristic to limit the repulsion list\n",
      "                double tTaken=0.0, int printInterval=10000,\n",
      "                double tot0=20.458, int nRepMax=0):\n",
      "  \n",
      "  cdef int nRest = len(restIndices)\n",
      "  cdef int nCoords = len(coords)\n",
      "    \n",
      "  if nCoords < 2:\n",
      "    raise NucCythonError('Too few coodinates')\n",
      "     \n",
      "  indices = set(restIndices.ravel())\n",
      "  if min(indices) < 0:\n",
      "    raise NucCythonError('Restraint index negative') \n",
      "  \n",
      "  if max(indices) >= nCoords:\n",
      "    data = (max(indices), nCoords)\n",
      "    raise NucCythonError('Restraint index \"%d\" out of bounds (> %d)' % data) \n",
      "  \n",
      "  if nCoords != len(masses):\n",
      "    raise NucCythonError('Masses list size does not match coordinates') \n",
      " \n",
      "  if nRest != len(restLimits):\n",
      "    raise NucCythonError('Number of restraint index pairs does not match number of restraint limits')   \n",
      "\n",
      "  if len(restAmbig) != nRest:\n",
      "    raise NucCythonError('Size of ambiguity list does not match number of restraints')   \n",
      "  \n",
      "  cdef int i, j, n, step, nViol, nRep = 0\n",
      "  \n",
      "  if not nRepMax:\n",
      "    nRepMax = nCoords*800\n",
      "    \n",
      "  cdef double d2, maxDelta, dx, dy, dz, ek, rmsd, tStep0, temp, fDist, fRep\n",
      "  cdef double deltaLim = 0.25 * repDist * repDist\n",
      "  cdef double Langevin_gamma\n",
      "\n",
      "  tStep0 = tStep * tot0 \n",
      "  beta /= tot0\n",
      "\n",
      "  veloc = numpy.random.normal(0.0, 1.0, (nCoords, 3))\n",
      "  veloc *= sqrt(tRef / getTemp(masses, veloc, nCoords))\n",
      "  \n",
      "  zeroArray = numpy.zeros((nCoords, 3))\n",
      "  repList = numpy.zeros((nRepMax, 2), numpy.int32)\n",
      "  \n",
      "  cdef ndarray[double, ndim=2] coordsPrev = numpy.array(coords)   \n",
      "  cdef ndarray[double, ndim=2] accel = numpy.array(zeroArray)    \n",
      "  cdef ndarray[double, ndim=2] forces\n",
      "  \n",
      "  t0 = time.time()\n",
      "\n",
      "  for step in range(nSteps):\n",
      "\n",
      "    if step == 0:\n",
      "      nRep = getRepulsionList(repList, coords, nCoords, nRepMax, repDist, radii) # Handle errors\n",
      "      \n",
      "      for i in range(nCoords):\n",
      "        coordsPrev[i,0] = coords[i,0]\n",
      "        coordsPrev[i,1] = coords[i,1]\n",
      "        coordsPrev[i,2] = coords[i,2]\n",
      "\n",
      "      forces = numpy.array(zeroArray)\n",
      "      fRep = getRepulsiveForce(repList, forces, coords, nRep,  fConstR, radii)\n",
      "      fDist = getRestraintForce(forces, coords, restIndices, restLimits, restAmbig, nRest, fConstD)\n",
      "\n",
      "      for i in range(nCoords):\n",
      "        accel[i,0] = forces[i,0] / masses[i]\n",
      "        accel[i,1] = forces[i,1] / masses[i]\n",
      "        accel[i,2] = forces[i,2] / masses[i]\n",
      "   \n",
      "    else:\n",
      "      maxDelta = 0.0\n",
      "      for i in range(nCoords): \n",
      "        dx = coords[i,0] - coordsPrev[i,0]\n",
      "        dy = coords[i,1] - coordsPrev[i,1]\n",
      "        dz = coords[i,2] - coordsPrev[i,2]\n",
      "        d2 = dx*dx + dy*dy + dz*dz\n",
      "        \n",
      "        if d2 > maxDelta:\n",
      "          maxDelta = d2\n",
      "          \n",
      "          if maxDelta > deltaLim:\n",
      "            break            \n",
      "\n",
      "      if maxDelta > deltaLim:\n",
      "        nRep = getRepulsionList(repList, coords, nCoords, nRepMax, repDist, radii) # Handle errors\n",
      "        \n",
      "        for i in range(nCoords):\n",
      "          coordsPrev[i,0] = coords[i,0]\n",
      "          coordsPrev[i,1] = coords[i,1]\n",
      "          coordsPrev[i,2] = coords[i,2]\n",
      " \n",
      "    updateMotion(masses, forces, accel, veloc, coords, nCoords, tRef, tStep0, beta)\n",
      "    \n",
      "    for i in range(nCoords):\n",
      "      forces[i,0] = 0.0\n",
      "      forces[i,1] = 0.0\n",
      "      forces[i,2] = 0.0\n",
      "      \n",
      "    fRep  = getRepulsiveForce(repList, forces, coords, nRep, fConstR, radii)\n",
      "    fDist = getRestraintForce(forces, coords, restIndices, restLimits, restAmbig, nRest, fConstD)\n",
      "\n",
      "    updateVelocity(masses, forces, accel, veloc, nCoords, tRef, tStep0,  beta)\n",
      "\n",
      "    if (printInterval > 0) and step % printInterval == 0:\n",
      "      temp = getTemp(masses, veloc, nCoords)\n",
      "      nViol, rmsd = getStats(restIndices, restLimits, coords, nRest)    \n",
      "      \n",
      "      data = (temp, fRep, fDist, rmsd, nViol, nRep)\n",
      "      print('temp:%7.2lf  fRep:%7.2lf  fDist:%7.2lf  rmsd:%7.2lf  nViol:%5d  nRep:%5d' % data)\n",
      "\n",
      "    tTaken += tStep\n",
      "  \n",
      "  return tTaken, nRep\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%cython\n",
      "from libc.math cimport exp, abs, sqrt, ceil\n",
      "from numpy cimport ndarray, double_t, int_t\n",
      "import numpy\n",
      "\n",
      "def calc_restraints(chromosomes, contact_dict, int particle_size=10000,\n",
      "                    float scale=1.0, float exponent=-0.33,\n",
      "                    float lower=0.8, float upper=1.2,\n",
      "                    int min_count=1):\n",
      "  \"\"\"\n",
      "  Function to convert single-cell contact data into distance restraints \n",
      "  for structure calculations.\n",
      "  \"\"\"\n",
      "    \n",
      "  cdef int i, j, k, a, b, nc, n, na, nb\n",
      "  cdef double dist\n",
      "  cdef ndarray[long, ndim=2] contacts      # Contact matrix (3:(posA, posB, nObs), nContacts)\n",
      "  cdef ndarray[double, ndim=2] restraints  # Distance restraints (6:(), nRestraints)\n",
      "  cdef ndarray[int, ndim=2] bin_matrix     # Temp array for binned contacts\n",
      "  cdef ndarray[int, ndim=1] seq_pos_a\n",
      "  cdef ndarray[int, ndim=1] seq_pos_b\n",
      "  cdef ndarray[int, ndim=2] limits         # shape: (chromoId, 2:[start, end])\n",
      "\n",
      "  num_contacts_dict = {} # Total num contacts for each chromosomes\n",
      "  pos_dict = {}          # Start sequence position for each particle in each chromosome\n",
      "  restraint_dict = {}    # Final restraints for each pair of chromosomes\n",
      "  \n",
      "  chromos = set(chromosomes)\n",
      "  nc = len(chromosomes)\n",
      "   \n",
      "  # Num contacts per chromo for mem allocations\n",
      "  for chrA in contact_dict:\n",
      "    if chrA not in chromos:\n",
      "      continue\n",
      "    \n",
      "    for chrB in contact_dict[chrA]:\n",
      "      if chrB not in chromos:\n",
      "        continue\n",
      "        \n",
      "      contacts = contact_dict[chrA][chrB]\n",
      "      n = len(contacts[0])\n",
      "      \n",
      "      if chrA in num_contacts_dict:\n",
      "        num_contacts_dict[chrA] += n\n",
      "      else:\n",
      "        num_contacts_dict[chrA] = n\n",
      "      \n",
      "      if chrB in num_contacts_dict:\n",
      "        num_contacts_dict[chrB] += n\n",
      "      else:\n",
      "        num_contacts_dict[chrB] = n\n",
      "        \n",
      "  chromo_idx = {chromo:i for i, chromo in enumerate(chromosomes)}\n",
      "  limits = numpy.zeros((nc,2), numpy.int32)\n",
      "  \n",
      "  # Get chromosome ranges\n",
      "  for chrA in contact_dict:\n",
      "    if chrA not in chromos:\n",
      "      continue\n",
      "\n",
      "    a = chromo_idx[chrA]\n",
      "    \n",
      "    for chrB in contact_dict[chrA]:\n",
      "      if chrB not in chromos:\n",
      "        continue\n",
      "        \n",
      "      b = chromo_idx[chrB]\n",
      "      contacts = contact_dict[chrA][chrB]\n",
      "      n = len(contacts[0])\n",
      "      \n",
      "      for i in range(n):\n",
      "        \n",
      "        if limits[a,0] == 0: # zero is not a valid seq pos anyhow\n",
      "          limits[a,0] = contacts[0,i]\n",
      "        elif contacts[0,i] < limits[a,0]:\n",
      "          limits[a,0] = contacts[0,i]\n",
      "        \n",
      "        if limits[b,0] == 0:\n",
      "          limits[b,0] = contacts[1,i]\n",
      "        elif contacts[1,i] < limits[b,0]:\n",
      "          limits[b,0] = contacts[1,i]\n",
      "        \n",
      "        if contacts[0,i] > limits[a,1]:\n",
      "          limits[a,1] = contacts[0,i]\n",
      "      \n",
      "        if contacts[1,i] > limits[b,1]:\n",
      "          limits[b,1] = contacts[1,i]\n",
      "\n",
      "            \n",
      "  # Shift extremities of chromosome ranges to lie exactly on particle region edges\n",
      "  for a in range(nc):\n",
      "    limits[a,0] = particle_size * (limits[a,0]/particle_size)\n",
      "    limits[a,1] = particle_size * <int>(ceil(<float>limits[a,1]/particle_size))\n",
      "\n",
      "  # Get particle region start positions for each chromosome  \n",
      "  for chromo in chromosomes:\n",
      "    a = chromo_idx[chromo]\n",
      "    pos_dict[chromo] = numpy.arange(limits[a,0], limits[a,1] + particle_size, particle_size, numpy.int32)  \n",
      "    \n",
      "  # Get restraint indices, do binning of contact observations\n",
      "  for chrA in contact_dict:\n",
      "    if chrA not in chromos:\n",
      "      continue\n",
      "      \n",
      "    a = chromo_idx[chrA]\n",
      "    na = len(pos_dict[chrA])\n",
      "    seq_pos_a = pos_dict[chrA]\n",
      "    restraint_dict[chrA] = {}\n",
      "    \n",
      "    for chrB in contact_dict[chrA]:\n",
      "      if chrB not in chromos:\n",
      "        continue\n",
      "        \n",
      "      b = chromo_idx[chrB]\n",
      "      nb = len(pos_dict[chrB])\n",
      "      seq_pos_b = pos_dict[chrB]\n",
      "      \n",
      "      contacts = contact_dict[chrA][chrB]\n",
      "      n = len(contacts[0])\n",
      "        \n",
      "      restraints = numpy.empty((6, n), float)\n",
      "      bin_matrix = numpy.zeros((na, nb), numpy.int32)\n",
      "\n",
      "      for i in range(n):\n",
      "        \n",
      "        # Find bin index for chromo A\n",
      "        for j in range(na):\n",
      "          if seq_pos_a[j] >= contacts[0,i]:\n",
      "            break\n",
      "        \n",
      "        else:\n",
      "          continue\n",
      "        \n",
      "        # Find bin index for chromo B\n",
      "        for k in range(nb):\n",
      "          if seq_pos_b[k] >= contacts[1,i]:\n",
      "            break\n",
      "            \n",
      "        else:\n",
      "          continue\n",
      "        \n",
      "        bin_matrix[j,k] += contacts[2,i]      \n",
      "        \n",
      "      #loop over all binned contacts, and calculate the constraint target distance\n",
      "      #using a powerlaw function and the number of observations\n",
      "      k = 0\n",
      "      for i in range(na):\n",
      "        for j in range(nb):\n",
      "          if bin_matrix[i,j] > 0:\n",
      "            if bin_matrix[i,j] < min_count:\n",
      "              continue\n",
      "              \n",
      "            dist = scale * bin_matrix[i,j] ** exponent\n",
      "            \n",
      "            restraints[0,k] = <double>i #binA\n",
      "            restraints[1,k] = <double>j #binB\n",
      "            restraints[2,k] = 1.0 # Weighting not currently used\n",
      "            restraints[3,k] = dist # target value\n",
      "            restraints[4,k] = dist * lower # constraint lower bound\n",
      "            restraints[5,k] = dist * upper # constraint upper bound\n",
      "            \n",
      "            k += 1\n",
      "\n",
      "      restraint_dict[chrA][chrB] = restraints[:,:k]\n",
      "        \n",
      "  return restraint_dict, pos_dict\n",
      "\n",
      "  \n",
      "def concatenate_restraints(restraint_dict, pos_dict, particle_size,\n",
      "                           backbone_lower=0.1, backbone_upper=1.1):\n",
      "  \"\"\"\n",
      "  Joins restraints stored in a dict by chromo pairs into long concatenated arrays.\n",
      "  Indices of restraints relate to concatenated chromo seq pos.\n",
      "  Add-in all the backbone restraints for sequential particles.\n",
      "  \"\"\"\n",
      "  \n",
      "  cdef int i, n, num_restraints, m\n",
      "  cdef int start_a, start_b\n",
      "  cdef float bbl = numpy.float32(backbone_lower)\n",
      "  cdef float bbu = numpy.float32(backbone_upper)\n",
      "  cdef double dist, size = particle_size\n",
      "  \n",
      "  # Get total max number restraints and final restraint index offset for all chromos    \n",
      "  i = 0\n",
      "  num_restraints = 0\n",
      "  chromo_idx_offset = {}\n",
      "  \n",
      "  for chr_a in sorted(pos_dict):\n",
      "    n = len(pos_dict[chr_a])\n",
      "    chromo_idx_offset[chr_a] = i\n",
      "    num_restraints += n-1 # One fewer because no link at end of chain\n",
      "    i += n\n",
      "        \n",
      "  for chr_a in restraint_dict:\n",
      "    for chr_b in restraint_dict[chr_a]:\n",
      "      num_restraints += restraint_dict[chr_a][chr_b].shape[1]\n",
      " \n",
      "  # Loop allocated arrays\n",
      "  cdef ndarray[int, ndim=1] positions\n",
      "  cdef ndarray[double, ndim=2] restraints\n",
      "  \n",
      "  # Final arrays which will hold identities of restrained particle pairs\n",
      "  # and the restraint distances for each\n",
      "  cdef ndarray[int, ndim=2] particle_indices = numpy.empty((num_restraints, 2), numpy.int32)\n",
      "  cdef ndarray[double, ndim=2] distances = numpy.empty((num_restraints, 2), float)\n",
      "\n",
      "  # Add backbone path restraints   \n",
      "  m = 0\n",
      "  for chr_a in pos_dict:\n",
      "    \n",
      "    positions = pos_dict[chr_a]\n",
      "    n = len(positions)\n",
      "    start_a = chromo_idx_offset[chr_a]\n",
      "    \n",
      "    for i in range(n-1):\n",
      "      \n",
      "      # Normally this is 1.0 for regular sized particles\n",
      "      dist = (positions[i+1]-positions[i])/size\n",
      "      \n",
      "      particle_indices[m,0] = i + start_a\n",
      "      particle_indices[m,1] = i + start_a + 1\n",
      "      \n",
      "      distances[m,0] = dist * bbl # lower\n",
      "      distances[m,1] = dist * bbu # upper\n",
      "      \n",
      "      m += 1\n",
      "      \n",
      "  # Add regular restraints for chromo pairs \n",
      "  for chr_a in restraint_dict:\n",
      "    start_a = chromo_idx_offset[chr_a] # Offset for chromo A particles\n",
      "   \n",
      "    for chr_b in restraint_dict[chr_a]:\n",
      "      start_b = chromo_idx_offset[chr_b] # Offset for chromo B particles\n",
      "      \n",
      "      restraints = restraint_dict[chr_a][chr_b]\n",
      "      n = restraints.shape[1]\n",
      "      \n",
      "      for i in range(n):\n",
      "        particle_indices[m,0] = <int>restraints[0,i] + start_a\n",
      "        particle_indices[m,1] = <int>restraints[1,i] + start_b\n",
      " \n",
      "        distances[m,0] = restraints[4,i] # lower\n",
      "        distances[m,1] = restraints[5,i] # upper\n",
      "        \n",
      "        m += 1\n",
      " \n",
      "  return particle_indices, distances\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "Simple annealing protocol"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def anneal_structure(coords, restraintIndices, restraintLimits,\n",
      "                    tempMax=5000.0, tempMin=10.0, tempSteps=500,\n",
      "                    dynSteps=100, timeStep=0.001, randSeed=None):\n",
      "  \"\"\"\n",
      "  A simple single-stage annealing protocol with exponential\n",
      "  temperature decay and sigmoidal repulsion scheme.\n",
      "  \"\"\"\n",
      "    \n",
      "  from random import seed\n",
      "  from math import atan, log, exp, pi\n",
      "  from numpy import array, random, ones, int32\n",
      "    \n",
      "  # Consider masses and fixed particles\n",
      "      \n",
      "  if randSeed is None:\n",
      "    seed()\n",
      "  else:\n",
      "    seed(randSeed)\n",
      "  \n",
      "  # Ensure inputs are numpy.ndarray  \n",
      "  coords = array(coords)\n",
      "  restraintIndices = array(restraintIndices, dtype=int32)\n",
      "  restraintLimits = array(restraintLimits)\n",
      "    \n",
      "  nPoints = len(coords)\n",
      "  \n",
      "  # Masses and radii not used, just set to 1.0\n",
      "  masses = radii = ones(nPoints, float)\n",
      "    \n",
      "  # Ambiguiuty not used, but represents the stride of sequential, grouped restraints\n",
      "  ambiguity = ones(len(restraintIndices), int32)\n",
      "\n",
      "  adj = 1.0 / atan(10.0)\n",
      "  decay = log(tempMax/tempMin)    \n",
      "\n",
      "  for step in range(tempSteps):\n",
      "    frac = step/float(tempSteps)\n",
      "    \n",
      "    # exponential temp decay\n",
      "    temp = tempMax * exp(-decay*frac)\n",
      "    \n",
      "    # sigmoidal repusion scheme\n",
      "    repulse = 0.5 + adj * atan(frac*20.0-10) / pi \n",
      "    \n",
      "    # update coordinates\n",
      "    runDynamics(coords, masses, radii, restraintIndices, restraintLimits,\n",
      "                ambiguity, temp, timeStep, dynSteps, repulse,\n",
      "                repDist=1.5, printInterval=int32(100))\n",
      "  \n",
      "    # Center\n",
      "    coords -= coords.mean(axis=0)\n",
      "  \n",
      "  return coords"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "Test annealing"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "from numpy import random\n",
      "\n",
      "# Random particle coordinates - 20 test points of (x,y,z) \n",
      "nCoords = 20\n",
      "coords = random.uniform(-1.0, 1.0, (nCoords,3)) \n",
      "  \n",
      "# Each pair represents a connection between a pair of particles\n",
      "# Restraint indices refer to indices in coordinate array\n",
      "rIndices = [(0, 1), (0, 5), (0, 18), (1, 0), (1, 2),\n",
      "            (1, 9), (2, 1), (2, 3), (2, 10), (3, 2),\n",
      "            (3, 4), (3, 17), (4, 3), (4, 5), (4, 11),\n",
      "            (5, 0), (5, 4), (5, 12), (6, 7), (6, 17),\n",
      "            (6, 19), (7, 6), (7, 14), (7, 15), (7, 16),\n",
      "            (8, 18), (9, 1), (10, 2), (11, 4), (12, 5),\n",
      "            (13, 17), (14, 7), (15, 7), (16, 7), (17, 3),\n",
      "            (17, 6), (17, 13), (18, 0), (18, 8), (19, 6)]\n",
      "  \n",
      "# Distances - upper and lower bounds for each particle\n",
      "# Unform upper and lower bounds for test restraints\n",
      "lower = 0.8\n",
      "upper = 1.2\n",
      "rDists = [(0.8, 1.2)] * len(rIndices)\n",
      "\n",
      "# Run calculation\n",
      "coords = anneal_structure(coords, rIndices, rDists)\n",
      "  \n",
      "# Show results\n",
      "for i, c in enumerate(coords):\n",
      "  print '%3d' % i, c\n",
      "  "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "Multi-scale repulsion scheme for whole genome structures"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def load_ncc_file(file_path):\n",
      "  \"\"\"Load chromosome and contact data from NCC format file, as output from NucProcess\"\"\"\n",
      "  \n",
      "  if file_path.endswith('.gz'):\n",
      "    import gzip\n",
      "    file_obj = gzip.open(file_path)\n",
      "  \n",
      "  else:\n",
      "    file_obj = open(file_path) \n",
      "  \n",
      "  # Observations are treated individually in single-cell Hi-C,\n",
      "  # i.e. no binning, so num_obs always 1 for each contact\n",
      "  num_obs = 1  \n",
      "    \n",
      "  contact_dict = {}\n",
      "  chromosomes = set()\n",
      "    \n",
      "  for line in file_obj:\n",
      "    chr_a, f_start_a, f_end_a, start_a, end_a, strand_a, chr_b, f_start_b, f_end_b, start_b, end_b, strand_b, ambig_group, pair_id, swap_pair = line.split()\n",
      "    \n",
      "    if strand_a == '+':\n",
      "      pos_a = int(f_start_a)\n",
      "    else:\n",
      "      pos_a = int(f_end_a)\n",
      "    \n",
      "    if strand_b == '+':\n",
      "      pos_b = int(f_start_b)       \n",
      "    else:\n",
      "      pos_b = int(f_end_b)\n",
      " \n",
      "    if chr_a > chr_b:\n",
      "      chr_a, chr_b = chr_b, chr_a\n",
      "      pos_a, pos_b = pos_b, pos_a\n",
      "    \n",
      "    if chr_a not in contact_dict:\n",
      "      contact_dict[chr_a] = {}\n",
      "      chromosomes.add(chr_a)\n",
      "      \n",
      "    if chr_b not in contact_dict[chr_a]:\n",
      "      contact_dict[chr_a][chr_b] = [] \n",
      "      chromosomes.add(chr_b)\n",
      "        \n",
      "    contact_dict[chr_a][chr_b].append((pos_a, pos_b, num_obs, int(ambig_group)))\n",
      "   \n",
      "  file_obj.close()\n",
      "  \n",
      "  chromo_limits = {}\n",
      "    \n",
      "  for chr_a in contact_dict:\n",
      "    for chr_b in contact_dict[chr_a]:\n",
      "      contacts = numpy.array(contact_dict[chr_a][chr_b]).T\n",
      "      contact_dict[chr_a][chr_b] = contacts\n",
      "      \n",
      "      seq_pos_a = contacts[1]\n",
      "      seq_pos_b = contacts[2]\n",
      "      \n",
      "      min_a = min(seq_pos_a)\n",
      "      max_a = max(seq_pos_a)\n",
      "      min_b = min(seq_pos_b)\n",
      "      max_b = max(seq_pos_b)\n",
      "        \n",
      "      if chr_a in chromo_limits:\n",
      "        prev_min, prev_max = chromo_limits[chr_a]\n",
      "        chromo_limits[chr_a] = [min(prev_min, min_a), max(prev_max, max_a)]\n",
      "      else:\n",
      "        chromo_limits[chr_a] = [min_a, max_a]\n",
      "      \n",
      "      if chr_b in chromo_limits:\n",
      "        prev_min, prev_max = chromo_limits[chr_b]\n",
      "        chromo_limits[chr_b] = [min(prev_min, min_b), max(prev_max, max_b)]\n",
      "      else:\n",
      "        chromo_limits[chr_b] = [min_b, max_b]\n",
      "         \n",
      "  chromosomes = sorted(chromosomes)      \n",
      "        \n",
      "  return chromosomes, chromo_limits, contact_dict\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def export_pdb_coords(file_path, coords_dict, seq_pos_dict, particle_size, scale=1.0, extended=True):\n",
      "  \n",
      "  from numpy import array, uint32, float32\n",
      "\n",
      "  alc = ' '\n",
      "  ins = ' '\n",
      "  prefix = 'HETATM'\n",
      "  line_format = '%-80.80s\\n'\n",
      "  \n",
      "  if extended:\n",
      "    pdb_format = '%-6.6s%5.1d %4.4s%s%3.3s %s%4.1d%s   %8.3f%8.3f%8.3f%6.2f%6.2f          %2.2s  %10d\\n'\n",
      "    ter_format = '%-6.6s%5.1d      %s %s%4.1d%s                                                     %10d\\n'\n",
      "  else:\n",
      "    pdb_format = '%-6.6s%5.1d %4.4s%s%3.3s %s%4.1d%s   %8.3f%8.3f%8.3f%6.2f%6.2f          %2.2s  \\n'\n",
      "    ter_format = '%-6.6s%5.1d      %s %s%4.1d%s                                                     \\n'\n",
      "\n",
      "  file_obj = open(file_path, 'w')\n",
      "  write = file_obj.write\n",
      "  \n",
      "  chromosomes = list(seq_pos_dict.keys())\n",
      "  num_models = len(coords_dict[chromosomes[0]])\n",
      "  title = 'NucDynamics genome structure export'\n",
      "  \n",
      "  write(line_format % 'TITLE     %s' % title)\n",
      "  write(line_format % 'REMARK 210') \n",
      "  write(line_format % 'REMARK 210 Atom type C is used for all particles')\n",
      "  write(line_format % 'REMARK 210 Atom number increases every %s bases' % particle_size)\n",
      "  write(line_format % 'REMARK 210 Residue code indicates chromosome')\n",
      "  write(line_format % 'REMARK 210 Residue number represents which sequence Mb the atom is in')\n",
      "  write(line_format % 'REMARK 210 Chain letter is different every chromosome, where Chr1=a, Chr2=b etc.')\n",
      "  \n",
      "  if extended:\n",
      "    file_obj.write(line_format % 'REMARK 210 Extended PDB format with particle seq. pos. in last column')\n",
      "  \n",
      "  file_obj.write(line_format % 'REMARK 210')\n",
      "  \n",
      "  pos_chromo = {}\n",
      "  \n",
      "  for m in range(num_models):\n",
      "    line = 'MODEL     %4d' % (m+1)\n",
      "    write(line_format  % line)\n",
      "    \n",
      "    c = 0\n",
      "    j = 1\n",
      "    seqPrev = None\n",
      "    for chromo in seq_pos_dict:\n",
      "      \n",
      "      if chromo.isdigit():\n",
      "        idx = int(chromo) - 1\n",
      "        chain_code = chr(ord('A')+idx)\n",
      "      \n",
      "      elif len(chromo) == 1:\n",
      "        chain_code = chromo.upper()\n",
      "        \n",
      "      else:\n",
      "        idx = chromosomes.index(chromo)\n",
      "        chain_code = chr(ord('a')+idx)\n",
      "      \n",
      "      tlc = chromo\n",
      "      while len(tlc) < 2:\n",
      "        tlc = '_' + tlc\n",
      "      \n",
      "      if len(tlc) == 2:\n",
      "        tlc = 'C' + tlc\n",
      "      \n",
      "      if len(tlc) > 3:\n",
      "        tlc = tlc[:3]\n",
      "      \n",
      "      chromo_model_coords = coords_dict[chromo][m]\n",
      "      \n",
      "      if not len(coords):\n",
      "        continue\n",
      "      \n",
      "      pos = seq_pos_dict[chromo]\n",
      "      \n",
      "      for i, seqPos in enumerate(pos):\n",
      "        c += 1\n",
      " \n",
      "        seqMb = int(seqPos//1e6) + 1\n",
      "        \n",
      "        if seqMb == seqPrev:\n",
      "          j += 1\n",
      "        else:\n",
      "          j = 1\n",
      "        \n",
      "        el = 'C'\n",
      "        a = 'C%d' % j\n",
      "          \n",
      "        aName = '%-3s' % a\n",
      "        x, y, z = chromo_model_coords[i] #XYZ coordinates\n",
      "         \n",
      "        seqPrev = seqMb\n",
      "        pos_chromo[c] = chromo\n",
      "        \n",
      "        if extended:\n",
      "          line  = pdb_format % (prefix,c,aName,alc,tlc,chain_code,seqMb,ins,x,y,z,0.0,0.0,el,seqPos)\n",
      "        else:\n",
      "          line  = pdb_format % (prefix,c,aName,alc,tlc,chain_code,seqMb,ins,x,y,z,0.0,0.0,el)\n",
      "          \n",
      "        write(line)\n",
      " \n",
      "    write(line_format  % 'ENDMDL')\n",
      " \n",
      "  for i in range(c-2):\n",
      "     if pos_chromo[i+1] == pos_chromo[i+2]:\n",
      "       line = 'CONECT%5.1d%5.1d' % (i+1, i+2)\n",
      "       write(line_format  % line)\n",
      " \n",
      "  write(line_format  % 'END')\n",
      "  file_obj.close()\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def remove_isolated_contacts(contact_dict, threshold=int(2e6)):\n",
      "  \"\"\"\n",
      "  Select only contacts which are within a given sequence separation of another\n",
      "  contact, for the same chromosome pair\n",
      "  \"\"\"\n",
      "    \n",
      "  for chromoA in contact_dict:\n",
      "    for chromoB in contact_dict[chromoA]:\n",
      "      contacts = contact_dict[chromoA][chromoB]\n",
      "      positions = numpy.array(contacts[:2], numpy.int32).T\n",
      "      \n",
      "      if len(positions): # Sometimes empty e.g. for MT, Y chromos \n",
      "        active_idx = getSupportedPairs(positions, numpy.int32(threshold))\n",
      "        contact_dict[chromoA][chromoB] = contacts[:,active_idx]\n",
      "     \n",
      "  return contact_dict\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def remove_violated_contacts(contact_dict, coords_dict, particle_seq_pos, particle_size, threshold=5.0):  \n",
      "  \"\"\"\n",
      "  Remove contacts whith structure distances that exceed a given threshold\n",
      "  \"\"\"\n",
      "  from numpy import int32\n",
      "    \n",
      "  for chr_a in contact_dict:\n",
      "    for chr_b in contact_dict[chr_a]:\n",
      "      contacts = contact_dict[chr_a][chr_b]\n",
      "      \n",
      "      contact_pos_a = contacts[0].astype(int32)\n",
      "      contact_pos_b = contacts[1].astype(int32)\n",
      "      \n",
      "      coords_a = coords_dict[chr_a]\n",
      "      coords_b = coords_dict[chr_b]\n",
      "\n",
      "      struc_dists = []\n",
      " \n",
      "      for m in range(len(coords_a)):\n",
      "        coord_data_a = getInterpolatedCoords([chr_a], {chr_a:contact_pos_a}, particle_seq_pos, coords_a[m])\n",
      "        coord_data_b = getInterpolatedCoords([chr_b], {chr_b:contact_pos_b}, particle_seq_pos, coords_b[m])\n",
      " \n",
      "        deltas = coord_data_a - coord_data_b\n",
      "        dists = numpy.sqrt((deltas*deltas).sum(axis=1))\n",
      "        struc_dists.append(dists)\n",
      "      \n",
      "      # Average over all conformational models\n",
      "      struc_dists = numpy.array(struc_dists).T.mean(axis=1)\n",
      "      \n",
      "      # Select contacts with distances below distance threshold\n",
      "      indices = (struc_dists < threshold).nonzero()[0]\n",
      "      contact_dict[chr_a][chr_b] = contacts[:,indices]\n",
      "        \n",
      "  return contact_dict\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_random_coords(pos_dict, chromosomes, num_models, radius=10.0):\n",
      "  \"\"\"\n",
      "  Get random, uniformly sampled coorinate positions, restricted to\n",
      "  a sphere of given radius\n",
      "  \"\"\"\n",
      "    \n",
      "  from random import uniform\n",
      "  from numpy import empty\n",
      "    \n",
      "  num_particles = sum([len(pos_dict[chromo]) for chromo in chromosomes])\n",
      "  coords = empty((num_models, num_particles, 3))\n",
      "  r2 = radius*radius\n",
      "    \n",
      "  for m in range(num_models):\n",
      "    \n",
      "    for i in range(num_particles):\n",
      "      x = y = z = radius\n",
      "\n",
      "      while x*x + y*y + z*z >= r2:\n",
      "        x = radius * (2*uniform(0,1) - 1)\n",
      "        y = radius * (2*uniform(0,1) - 1)\n",
      "        z = radius * (2*uniform(0,1) - 1)\n",
      "      \n",
      "      coords[m,i] = [x,y,z]\n",
      "\n",
      "  return coords"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def pack_chromo_coords(coords_dict, chromosomes):\n",
      "  \"\"\"\n",
      "  Place chromosome 3D coordinates stored in a dictionary keyed by\n",
      "  chromosome name into a single, ordered array. The chromosomes argument\n",
      "  is required to set the correct array storage order.\n",
      "  \"\"\"\n",
      "  \n",
      "  from numpy import empty\n",
      "\n",
      "  chromo_num_particles = [len(coords_dict[chromo][0]) for chromo in chromosomes]\n",
      "  n_particles = sum(chromo_num_particles)\n",
      "  n_models = len(coords_dict[chromosomes[0]])  \n",
      "  coords = empty((n_models, n_particles, 3), float)\n",
      "  \n",
      "  j = 0\n",
      "  for i, chromo in enumerate(chromosomes):\n",
      "    span = chromo_num_particles[i]\n",
      "    coords[:,j:j+span] = coords_dict[chromo]\n",
      "    j += span\n",
      "      \n",
      "  return coords"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def unpack_chromo_coords(coords, chromosomes, seq_pos_dict):\n",
      "  \"\"\"\n",
      "  Exctract coords for multiple chromosomes stored in a single array into\n",
      "  a dictionary, keyed by chromosome name. The chromosomes argument is required\n",
      "  to get the correct array storage order.\n",
      "  \"\"\"\n",
      "\n",
      "  chromo_num_particles = [len(seq_pos_dict[chromo]) for chromo in chromosomes]\n",
      "  n_seq_pos = sum(chromo_num_particles)\n",
      "  n_models, n_particles, dims = coords.shape\n",
      "\n",
      "  if n_seq_pos != n_particles:\n",
      "    msg = 'Model coordinates must be an array of num models x %d' % (n_seq_pos,)\n",
      "    raise(Exception(msg))  \n",
      "  \n",
      "  coords_dict = {}\n",
      "        \n",
      "  j = 0\n",
      "  for i, chromo in enumerate(chromosomes):\n",
      "    span = chromo_num_particles[i]\n",
      "    coords_dict[chromo] = coords[:,j:j+span] # all models, slice\n",
      "    j += span\n",
      " \n",
      "  return coords_dict"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def anneal_genome(chromosomes, contact_dict, num_models, particle_size,\n",
      "                  general_calc_params, anneal_params,\n",
      "                  prev_seq_pos_dict=None, start_coords=None):\n",
      "    \"\"\"\n",
      "    Use chromosome contact data to generate distance restraints and then\n",
      "    apply a simulated annealing protocul to generate/refine coordinates.\n",
      "    Starting coordinates may be random of from a previous (e.g. lower\n",
      "    resolution) stage.\n",
      "    \"\"\"\n",
      "    \n",
      "    from numpy import int32, ones, empty\n",
      "    from math import log, exp, atan, pi\n",
      "    import gc\n",
      "    \n",
      "    random.seed(general_calc_params['random_seed'])\n",
      "    particle_size = int32(particle_size)\n",
      "    \n",
      "    # Calculate distance restrains from contact data   \n",
      "    restraint_dict, seq_pos_dict = calc_restraints(chromosomes, contact_dict, particle_size,\n",
      "                                                   scale=1.0, exponent=general_calc_params['dist_power_law'],\n",
      "                                                   lower=general_calc_params['contact_dist_lower'],\n",
      "                                                   upper=general_calc_params['contact_dist_upper'])\n",
      "    \n",
      "    # Concatenate chromosomal data into a single array of particle restraints\n",
      "    # for structure calculation. Add backbone restraints between seq. adjasent particles.\n",
      "    restraint_indices, restraint_dists = concatenate_restraints(restraint_dict, seq_pos_dict, particle_size,\n",
      "                                                                general_calc_params['backbone_dist_lower'],\n",
      "                                                                general_calc_params['backbone_dist_upper'])\n",
      " \n",
      "    # Setup starting structure\n",
      "    if (start_coords is None) or (prev_seq_pos_dict is None):\n",
      "      coords = get_random_coords(seq_pos_dict, chromosomes, num_models,\n",
      "                                 general_calc_params['random_radius'])\n",
      "      \n",
      "      num_coords = coords.shape[1]\n",
      "            \n",
      "    else:\n",
      "      # Convert starting coord dict into single array\n",
      "      coords = pack_chromo_coords(start_coords, chromosomes)\n",
      "      num_coords = sum([len(seq_pos_dict[c]) for c in chromosomes])\n",
      "        \n",
      "      if coords.shape[1] != num_coords: # Change of particle_size\n",
      "        interp_coords = empty((num_models, num_coords, 3))\n",
      "        \n",
      "        for m in range(num_models): # Starting coords interpolated from previous particle positions\n",
      "          interp_coords[m] = getInterpolatedCoords(chromosomes, seq_pos_dict, prev_seq_pos_dict, coords[m])\n",
      "        \n",
      "        coords = interp_coords\n",
      "        \n",
      "    # Equal unit masses and radii for all particles\n",
      "    masses = ones(num_coords,  float)\n",
      "    radii = ones(num_coords,  float)\n",
      "    \n",
      "    # Ambiguiity strides not used here, so set to 1\n",
      "    num_restraints = len(restraint_indices)\n",
      "    ambiguity = ones(num_restraints,  int32)\n",
      "        \n",
      "    # Below will be set to restrict memory allocation in the repusion list\n",
      "    # (otherwise all vs all can be huge)\n",
      "    n_rep_max = int32(0)\n",
      "    \n",
      "    # Annealing parameters\n",
      "    temp_start = anneal_params['temp_start']\n",
      "    temp_end = anneal_params['temp_end']\n",
      "    temp_steps = anneal_params['temp_steps']\n",
      "    \n",
      "    # Setup annealig schedule: setup temps and repulsive terms\n",
      "    adj = 1.0 / atan(10.0)\n",
      "    decay = log(temp_start/temp_end)    \n",
      "    anneal_schedule = []\n",
      "    \n",
      "    for step in range(temp_steps):\n",
      "      frac = step/float(temp_steps)\n",
      "    \n",
      "      # exponential temp decay\n",
      "      temp = temp_start * exp(-decay*frac)\n",
      "    \n",
      "      # sigmoidal repusion scheme\n",
      "      repulse = 0.5 + adj * atan(frac*20.0-10) / pi \n",
      "      \n",
      "      anneal_schedule.append((temp, repulse))  \n",
      "        \n",
      "    # Paricle dynamics parameters\n",
      "    # (these need not be fixed for all stages, but are for simplicity)    \n",
      "    dyn_steps = anneal_params['dynamics_steps']\n",
      "    time_step = anneal_params['time_step']\n",
      "       \n",
      "    # Update coordinates in the annealing schedule\n",
      "    time_taken = 0.0\n",
      "    \n",
      "    for m in range(num_models): # For each repeat calculation\n",
      "      model_coords = coords[m]\n",
      "        \n",
      "      for temp, repulse in anneal_schedule:\n",
      "        gc.collect() # Try to free some memory\n",
      "        \n",
      "        # Update coordinates for this temp\n",
      "        dt, n_rep_max = runDynamics(model_coords, masses, radii, restraint_indices, restraint_dists,\n",
      "                                    ambiguity, temp, time_step, dyn_steps, repulse, nRepMax=n_rep_max)\n",
      "      \n",
      "        n_rep_max = int32(1.05 * n_rep_max) # Base on num in prev cycle, plus an overhead\n",
      "        time_taken += dt\n",
      "  \n",
      "      # Center\n",
      "      model_coords -= model_coords.mean(axis=0)\n",
      "      \n",
      "      # Update\n",
      "      coords[m] = model_coords\n",
      "    \n",
      "    # Convert from single coord array to dict keyed by chromosome\n",
      "    coords_dict = unpack_chromo_coords(coords, chromosomes, seq_pos_dict)\n",
      "    \n",
      "    return coords_dict, seq_pos_dict\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from time import time\n",
      "\n",
      "# Number of alternative conformations to generate from repeat calculations\n",
      "# with different random starting coordinates\n",
      "num_models = 2  \n",
      "\n",
      "# Parameters to setup restraints and starting coords\n",
      "general_calc_params = {'dist_power_law':-0.33, \n",
      "                       'contact_dist_lower':0.8, 'contact_dist_upper':1.2,\n",
      "                       'backbone_dist_lower':0.1, 'backbone_dist_upper':1.1,\n",
      "                       'random_seed':int(time()), 'random_radius':10.0}\n",
      "\n",
      "# Annealing & dyamics parameters: the same for all stages\n",
      "# (this is cautious, but not an absolute requirement)\n",
      "anneal_params = {'temp_start':5000.0, 'temp_end':10.0, 'temp_steps':500,\n",
      "                 'dynamics_steps':100, 'time_step':0.001}\n",
      "\n",
      "# Hierarchical scale protocol\n",
      "particle_sizes = [8e6, 4e6, 2e6, 4e5, 2e5, 1e5]\n",
      "\n",
      "# Load single-cell Hi-C data from NCC contact file, as output from NucProcess\n",
      "chromosomes, chromo_limits, contact_dict = load_ncc_file('example_chromo_data/Cell_1_contacts.ncc')\n",
      "\n",
      "# Only use contacts which are supported by others nearby in sequence, in the initial instance\n",
      "remove_isolated_contacts(contact_dict, threshold=2e6)\n",
      "\n",
      "# Initial coords will be random\n",
      "start_coords = None\n",
      "\n",
      "# Record partile positions from previous stages\n",
      "# so that coordinates can be interpolated to higher resolution\n",
      "prev_seq_pos = None\n",
      "\n",
      "for stage, particle_size in enumerate(particle_sizes):\n",
      "    \n",
      "    print \"Running structure caculation stage %d (%d kb)\" % (stage+1, (particle_size/1e3))\n",
      "    \n",
      "    # Can remove large violations (noise contacts inconsistent with structure)\n",
      "    # once we have a resonable resolution structure\n",
      "    if stage == 4:\n",
      "        remove_violated_contacts(contact_dict, coords_dict, particle_seq_pos,\n",
      "                                 particle_size, threshold=6.0)\n",
      "    elif stage == 5:\n",
      "        remove_violated_contacts(contact_dict, coords_dict, particle_seq_pos,\n",
      "                                 particle_size, threshold=5.0)\n",
      "    \n",
      "    coords_dict, particle_seq_pos = anneal_genome(chromosomes, contact_dict, num_models, particle_size,\n",
      "                                                  general_calc_params, anneal_params,\n",
      "                                                  prev_seq_pos, start_coords)\n",
      "    \n",
      "    # Next stage based on previous stage's 3D coords\n",
      "    # and thier respective seq. positions\n",
      "    start_coords = coords_dict\n",
      "    prev_seq_pos = particle_seq_pos\n",
      "\n",
      "# Save final coords as PDB format file\n",
      "save_path = 'example_chromo_data/Cell_1_structure.pdb'\n",
      "export_pdb_coords(save_path, coords_dict, particle_seq_pos, particle_size)\n",
      "print('Saved structure file to: %s' % save_path)\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}